File: dfs/islands/number_of_islands.rb
# Approach 1: DFS
# @param {Character[][]} grid
# @return {Integer}
def num_islands(grid)
    return 0 if grid.empty? || grid[0].empty?
    m, n, count = grid.size, grid[0].size, 0

    0.upto(m - 1) do |i|
        0.upto(n - 1) do |j|
            next if grid[i][j] == '0'
            dfs(grid, i, j)
            count += 1
        end
    end

    count
end

def dfs(grid, i, j)
    m, n = grid.size, grid[0].size
    return if grid[i][j] == '0' # Optional as we're limiting dfs to only when unvisited (== '1')
    grid[i][j] = '0'    # Sink the island/mark as visited
    dirs = [[-1, 0], [1, 0], [0, -1], [0, 1]]

    dirs.each do |di, dj|
        x, y = i + di, j + dj
        next if !x.between?(0, m - 1) || !y.between?(0, n - 1) || grid[x][y] == '0'
        dfs(grid, x, y)
    end
end

# Approach 2: BFS
# @param {Character[][]} grid
# @return {Integer}
def num_islands_bfs(grid)
    return 0 if grid.empty? || grid[0].empty?
    m, n, count = grid.size, grid[0].size, 0

    0.upto(m - 1) do |i|
        0.upto(n - 1) do |j|
            next if grid[i][j] == '0'
            bfs(grid, i, j)
            count += 1
        end
    end

    count
end

def bfs(grid, i, j)
    m, n, q = grid.size, grid[0].size, [[i, j]]
    grid[i][j] = '0' # Mark as visited
    dirs = [[-1, 0], [1, 0], [0, -1], [0, 1]]

    while !q.empty?
        i, j = q.shift()

        dirs.each do |di, dj|
            x, y = i + di, j + dj
            next if !x.between?(0, m - 1) || !y.between?(0, n - 1) || grid[x][y] == '0'
            q.push([x, y])
            grid[x][y] = '0'
        end
    end
end


class DisjoinSet
    attr_accessor :parent,  :rank, :count

    def initialize()
        @parent = {}
        @rank = {}
        @count = 0
    end

    def set_parent(x)
        @parent[x] = x
        @rank[x] = 0
        @count += 1
    end

    def find(x) # with path compression
        @parent[x] = find(@parent[x]) if x != @parent[x]
        @parent[x]
    end

    def union(x, y)
        par_x, par_y = find(x), find(y)
        return if par_x == par_y

        # Union by rank means make the element with higher rank the parent
        # to keep the depth of the tree to a minimum
        if @rank[par_x] > @rank[par_y]      # par_x has greater rank,
            @parent[par_y] = par_x          # So par_x becomes parent of par_y
        elsif @rank[par_y] > @rank[par_x]   # par_y has greater rank,
            @parent[par_x] = par_y          # So par_y becomes parent of par_x
        else                                # Same rank,
            @parent[par_y] = par_x          # so make one parent of the other
            @rank[par_x] += 1               # only increase rank of parent when ranks are same
        end
        @count -= 1                         # Total count of sets reduces by as two sets become one
    end
end

# Approach 3: UF
# @param {Character[][]} grid
# @return {Integer}
def num_islands_uf(grid)
    return 0 if grid.empty? || grid[0].empty?
    m, n = grid.size, grid[0].size
    ds = DisjoinSet.new()
    dirs = [[-1, 0], [1, 0], [0, -1], [0, 1]]

    0.upto(m - 1) do |i|
        0.upto(n - 1) do |j|
            next if grid[i][j] == '0'
            ds.set_parent([i, j])
            dirs.each do |di, dj|
                x, y = i + di, j + dj
                next if !x.between?(0, m - 1) || !y.between?(0, n - 1) || !ds.parent.key?([x, y])
                ds.union([i, j], [x, y]) # Union takes constant time, due to union by rank and path compression
            end
        end
    end

    ds.count
end

# 200. Number of Islands
# https://leetcode.com/problems/number-of-islands/description/


# Approach 1: DFS
# Time: O(m * n)
# Space: O(m * n) in case that the grid map is filled with lands where DFS goes by m * n deep

# Approach 2: BFS
# Time: O(m * n)
# Space: O(min(m, n))

# Approach 3: Union-Find
# Time: O(m * n)
# Space: O(m * n)

require 'test/unit'
extend Test::Unit::Assertions

grid = ["11110",
        "11010",
        "11000",
        "00000"]
grid2 = ['100',
         '000',
         '001']
grid3 = ["111111",
         "100001",
         "101101",
         "100001",
         "111111"]
grid4 = [["1","1","1","1","1","0","1","1","1","1","1","1","1","1","1","0","1","0","1","1"],
        ["0","1","1","1","1","1","1","1","1","1","1","1","1","0","1","1","1","1","1","0"],
        ["1","0","1","1","1","0","0","1","1","0","1","1","1","1","1","1","1","1","1","1"],
        ["1","1","1","1","0","1","1","1","1","1","1","1","1","1","1","1","1","1","1","1"],
        ["1","0","0","1","1","1","1","1","1","1","1","1","1","1","1","1","1","1","1","1"],
        ["1","0","1","1","1","1","1","1","0","1","1","1","0","1","1","1","0","1","1","1"],
        ["0","1","1","1","1","1","1","1","1","1","1","1","0","1","1","0","1","1","1","1"],
        ["1","1","1","1","1","1","1","1","1","1","1","1","0","1","1","1","1","0","1","1"],
        ["1","1","1","1","1","1","1","1","1","1","0","1","1","1","1","1","1","1","1","1"],
        ["1","1","1","1","1","1","1","1","1","1","1","1","1","1","1","1","1","1","1","1"],
        ["0","1","1","1","1","1","1","1","0","1","1","1","1","1","1","1","1","1","1","1"],
        ["1","1","1","1","1","1","1","1","1","1","1","1","1","1","1","1","1","1","1","1"],
        ["1","1","1","1","1","1","1","1","1","1","1","1","1","1","1","1","1","1","1","1"],
        ["1","1","1","1","1","0","1","1","1","1","1","1","1","0","1","1","1","1","1","1"],
        ["1","0","1","1","1","1","1","0","1","1","1","0","1","1","1","1","0","1","1","1"],
        ["1","1","1","1","1","1","1","1","1","1","1","1","0","1","1","1","1","1","1","0"],
        ["1","1","1","1","1","1","1","1","1","1","1","1","1","0","1","1","1","1","0","0"],
        ["1","1","1","1","1","1","1","1","1","1","1","1","1","1","1","1","1","1","1","1"],
        ["1","1","1","1","1","1","1","1","1","1","1","1","1","1","1","1","1","1","1","1"],
        ["1","1","1","1","1","1","1","1","1","1","1","1","1","1","1","1","1","1","1","1"]]

assert_equal(num_islands(grid), 1)
assert_equal(num_islands(grid2), 2)
assert_equal(num_islands(grid3), 2)
assert_equal(num_islands(grid4), 1)

grid = ["11110",
        "11010",
        "11000",
        "00000"]
grid2 = ['100',
         '000',
         '001']
grid3 = ["111111",
         "100001",
         "101101",
         "100001",
         "111111"]
grid4 = [["1","1","1","1","1","0","1","1","1","1","1","1","1","1","1","0","1","0","1","1"],
        ["0","1","1","1","1","1","1","1","1","1","1","1","1","0","1","1","1","1","1","0"],
        ["1","0","1","1","1","0","0","1","1","0","1","1","1","1","1","1","1","1","1","1"],
        ["1","1","1","1","0","1","1","1","1","1","1","1","1","1","1","1","1","1","1","1"],
        ["1","0","0","1","1","1","1","1","1","1","1","1","1","1","1","1","1","1","1","1"],
        ["1","0","1","1","1","1","1","1","0","1","1","1","0","1","1","1","0","1","1","1"],
        ["0","1","1","1","1","1","1","1","1","1","1","1","0","1","1","0","1","1","1","1"],
        ["1","1","1","1","1","1","1","1","1","1","1","1","0","1","1","1","1","0","1","1"],
        ["1","1","1","1","1","1","1","1","1","1","0","1","1","1","1","1","1","1","1","1"],
        ["1","1","1","1","1","1","1","1","1","1","1","1","1","1","1","1","1","1","1","1"],
        ["0","1","1","1","1","1","1","1","0","1","1","1","1","1","1","1","1","1","1","1"],
        ["1","1","1","1","1","1","1","1","1","1","1","1","1","1","1","1","1","1","1","1"],
        ["1","1","1","1","1","1","1","1","1","1","1","1","1","1","1","1","1","1","1","1"],
        ["1","1","1","1","1","0","1","1","1","1","1","1","1","0","1","1","1","1","1","1"],
        ["1","0","1","1","1","1","1","0","1","1","1","0","1","1","1","1","0","1","1","1"],
        ["1","1","1","1","1","1","1","1","1","1","1","1","0","1","1","1","1","1","1","0"],
        ["1","1","1","1","1","1","1","1","1","1","1","1","1","0","1","1","1","1","0","0"],
        ["1","1","1","1","1","1","1","1","1","1","1","1","1","1","1","1","1","1","1","1"],
        ["1","1","1","1","1","1","1","1","1","1","1","1","1","1","1","1","1","1","1","1"],
        ["1","1","1","1","1","1","1","1","1","1","1","1","1","1","1","1","1","1","1","1"]]

assert_equal(num_islands_bfs(grid), 1)
assert_equal(num_islands_bfs(grid2), 2)
assert_equal(num_islands_bfs(grid3), 2)
assert_equal(num_islands_bfs(grid4), 1)

grid = ["11110",
        "11010",
        "11000",
        "00000"]
grid2 = ['100',
         '000',
         '001']
grid3 = ["111111",
         "100001",
         "101101",
         "100001",
         "111111"]
grid4 = [
    ["1","1","1","1","1","0","1","1","1","1","1","1","1","1","1","0","1","0","1","1"],
    ["0","1","1","1","1","1","1","1","1","1","1","1","1","0","1","1","1","1","1","0"],
    ["1","0","1","1","1","0","0","1","1","0","1","1","1","1","1","1","1","1","1","1"],
    ["1","1","1","1","0","1","1","1","1","1","1","1","1","1","1","1","1","1","1","1"],
    ["1","0","0","1","1","1","1","1","1","1","1","1","1","1","1","1","1","1","1","1"],
    ["1","0","1","1","1","1","1","1","0","1","1","1","0","1","1","1","0","1","1","1"],
    ["0","1","1","1","1","1","1","1","1","1","1","1","0","1","1","0","1","1","1","1"],
    ["1","1","1","1","1","1","1","1","1","1","1","1","0","1","1","1","1","0","1","1"],
    ["1","1","1","1","1","1","1","1","1","1","0","1","1","1","1","1","1","1","1","1"],
    ["1","1","1","1","1","1","1","1","1","1","1","1","1","1","1","1","1","1","1","1"],
    ["0","1","1","1","1","1","1","1","0","1","1","1","1","1","1","1","1","1","1","1"],
    ["1","1","1","1","1","1","1","1","1","1","1","1","1","1","1","1","1","1","1","1"],
    ["1","1","1","1","1","1","1","1","1","1","1","1","1","1","1","1","1","1","1","1"],
    ["1","1","1","1","1","0","1","1","1","1","1","1","1","0","1","1","1","1","1","1"],
    ["1","0","1","1","1","1","1","0","1","1","1","0","1","1","1","1","0","1","1","1"],
    ["1","1","1","1","1","1","1","1","1","1","1","1","0","1","1","1","1","1","1","0"],
    ["1","1","1","1","1","1","1","1","1","1","1","1","1","0","1","1","1","1","0","0"],
    ["1","1","1","1","1","1","1","1","1","1","1","1","1","1","1","1","1","1","1","1"],
    ["1","1","1","1","1","1","1","1","1","1","1","1","1","1","1","1","1","1","1","1"],
    ["1","1","1","1","1","1","1","1","1","1","1","1","1","1","1","1","1","1","1","1"]]

assert_equal(num_islands_uf(grid), 1)
assert_equal(num_islands_uf(grid2), 2)
assert_equal(num_islands_uf(grid3), 2)
assert_equal(num_islands_uf(grid4), 1)

File: dfs/islands/number_of_islands.py
from typing import List
import collections


class Solution:
    def numIslands(self, grid: List[List[str]]) -> int:
        if not grid or not grid[0]: return 0
        m, n = len(grid), len(grid[0])
        count = 0

        for i in range(m):
            for j in range(n):
                if grid[i][j] == '1':
                    self.bfs(grid, i, j)
                    count += 1

        return count

    def bfs(self, grid, r, c):
        queue = collections.deque()
        queue.append((r, c))
        grid[r][c] = '0'

        while queue:
            directions = [(0,1), (0,-1), (-1,0), (1,0)]
            r, c = queue.popleft()

            for d in directions:
                nr, nc = r + d[0], c + d[1]
                if self.is_valid(grid, nr, nc) and grid[nr][nc] == '1':
                    queue.append((nr, nc))
                    grid[nr][nc] = '0'

    def is_valid(self, grid, r, c):
        m, n = len(grid), len(grid[0])
        if r < 0 or c < 0 or r >= m or c >= n:
            return False
        return True

class DS:
    def __init__(self):
        self.parent = {}
        self.rank = {}
        self.count = 0

    def find(self, x):
        if x != self.parent[x]: self.parent[x] = self.find(self.parent[x])
        return self.parent[x]

    def has_parent(self, x):
        return x in self.parent

    def set_parent(self, x):
        self.parent[x] = x
        self.rank[x] = 0
        self.count += 1

    def union(self, x, y):
        px, py = self.find(x), self.find(y)
        if px == py: return False

        if self.rank[px] > self.rank[py]:
            self.parent[py] = px
        elif self.rank[py] > self.rank[px]:
            self.parent[px] = py
        else:
            self.parent[py] = px
            self.rank[px] += 1

        self.count -= 1
        return True

    def print_ds(self):
        print(self.parent)
        print(self.count)


class Solution:
    def numIslands(self, grid: List[List[str]]) -> int:
        if not grid or not grid[0]: return 0
        m, n, ds = len(grid), len(grid[0]), DS()
        dirs = [(-1, 0), (1, 0), (0, -1), (0, 1)]

        for i in range(m):
            for j in range(n):
                if grid[i][j] == "0": continue
                if not ds.has_parent((i, j)): ds.set_parent((i, j))

                for dx, dy in dirs:
                    x, y = i + dx, j + dy
                    if not self.valid_land(grid, m, n, x, y): continue
                    if not ds.has_parent((x, y)): ds.set_parent((x, y))
                    ds.union((i, j), (x, y))

        return ds.count

    def valid_land(self, grid, m, n, r, c):
        return 0 <= r <= m - 1 and 0 <= c <= n - 1 and grid[r][c] == "1"


class Solution2:
    def numIslands(self, grid: List[List[str]]) -> int:
        if not grid or not grid[0]: return 0
        m, n = len(grid), len(grid[0])

        count = 0
        for i in range(m):
            for j in range(n):
                if grid[i][j] == '0': continue
                self.dfs(grid, m, n, i, j)
                count += 1

        return count

    def dfs(self, grid, m, n, i, j):
        grid[i][j] = '0'
        dirs = [(-1, 0), (1, 0), (0, -1), (0, 1)]
        for d in dirs:
            dx, dy = d
            x, y = i + dx, j + dy
            if not self.valid_land(grid, m, n, x, y): continue
            self.dfs(grid, m, n, x, y)

    def valid_land(self, grid, m, n, r, c):
        return 0 <= r <= m - 1 and 0 <= c <= n - 1 and grid[r][c] == "1"



# 200. Number of Islands
# https://leetcode.com/problems/number-of-islands/description/


sol = Solution()

assert sol.numIslands(
    [["1","1","1","1","0"],
     ["1","1","0","1","0"],
     ["1","1","0","0","0"],
     ["0","0","0","0","0"]]) == 1


    #  ["0","0","0","0","0"]
    #  ["0","0","0","0","0"]
    #  ["0","0","0","0","0"]
    #  ["0","0","0","0","0"]



# dfs(0, 0, 0) [0]
#     dfs(1, 0, 1) [0, 1]
#         dfs(1,1,2) [0, 1, 2]
#             dfs(1,3,2) [0 1 2 2]
#             [0 1 2 2 0]

# (r+1, c, 1)
# (r-1, c, 2)
# (r, c+1, 3)
# (r, c-1, 4)

# [[1,1,0],
#  [0,1,1],
#  [0,0,0],
#  [1,1,1],
#  [0,1,0]]


# 0 3 1 3 0 0 0 0

# 0 3 1 0 3

# - -
#  |